File: frontend/src/types/catalog.ts
import http from "../http";
import { BaseResource } from "./base";

export interface Catalog {
  id: number;
  name: string;
  url: string;
  sgbd: string;
  username: string;
  password: string;
}

export type CatalogCreate = Omit<Catalog, "id">;
export type CatalogUpdate = Partial<Omit<Catalog, "id">>;

export interface CatalogConnectionStatus {
  catalog_id: number;
  state: "ok" | "error";
  message: string;
}

export interface CatalogSchemaEntry {
  id: number;
  catalog_id: number;
  schema_name: string;
  table_name: string | null;
  column_name: string | null;
  manually_added: boolean;
  present_in_database: boolean;
}

export interface CatalogSchemaEntryCreate {
  schema_name: string;
  table_name?: string | null;
  column_name?: string | null;
}

export interface CatalogSchemaSyncResult {
  catalog_id: number;
  schemas: number;
  tables: number;
  columns: number;
  created_entries: number;
  deleted_entries: number;
  deleted_rules: number;
}

export interface CatalogTablePreviewRequest {
  user_id: number;
  schema_name: string;
  table_name: string;
  limit?: number;
}

export interface CatalogPreviewDataset {
  columns: string[];
  rows: Record<string, unknown>[];
  error?: string | null;
}

export interface CatalogTablePreview {
  catalog_id: number;
  schema_name: string;
  table_name: string;
  limit: number;
  before_maskql: CatalogPreviewDataset;
  after_maskql: CatalogPreviewDataset;
}

export class Catalogs extends BaseResource<
  Catalog,
  CatalogCreate,
  CatalogUpdate
> {
  constructor() {
    super("/catalogs");
  }

  async listStatuses(): Promise<CatalogConnectionStatus[]> {
    const { data } = await http.get<CatalogConnectionStatus[]>(
      `${this.endpoint}/status`,
    );
    return data;
  }

  async listSchema(catalogId: number): Promise<CatalogSchemaEntry[]> {
    const { data } = await http.get<CatalogSchemaEntry[]>(
      `${this.endpoint}/${catalogId}/schema`,
    );
    return data;
  }

  async createSchemaEntry(
    catalogId: number,
    payload: CatalogSchemaEntryCreate,
  ): Promise<CatalogSchemaEntry> {
    const { data } = await http.post<CatalogSchemaEntry>(
      `${this.endpoint}/${catalogId}/schema`,
      payload,
    );
    return data;
  }

  async deleteSchemaEntry(catalogId: number, entryId: number): Promise<void> {
    await http.delete(`${this.endpoint}/${catalogId}/schema/${entryId}`);
  }

  async syncSchema(catalogId: number): Promise<CatalogSchemaSyncResult> {
    const { data } = await http.post<CatalogSchemaSyncResult>(
      `${this.endpoint}/${catalogId}/schema/sync`,
    );
    return data;
  }

  async previewTable(
    catalogId: number,
    payload: CatalogTablePreviewRequest,
  ): Promise<CatalogTablePreview> {
    const { data } = await http.post<CatalogTablePreview>(
      `${this.endpoint}/${catalogId}/schema/preview`,
      payload,
    );
    return data;
  }
}

export const CatalogAPI = new Catalogs();

File: maskql/schemas/catalog.py
from __future__ import annotations
from typing import Any, Optional
from pydantic import field_validator
from sqlmodel import SQLModel

class CatalogCreate(SQLModel):
    name: str
    url: str
    sgbd: str
    username: str
    password: str

class CatalogPatch(SQLModel):
    name: Optional[str] = None
    url: Optional[str] = None
    sgbd: Optional[str] = None
    username: Optional[str] = None
    password: Optional[str] = None

class CatalogRead(SQLModel):
    id: int
    name: str
    url: str
    sgbd: str
    username: str


class CatalogConnectionStatusRead(SQLModel):
    catalog_id: int
    state: str
    message: str


class CatalogSchemaEntryRead(SQLModel):
    id: int
    catalog_id: int
    schema_name: str
    table_name: Optional[str] = None
    column_name: Optional[str] = None
    manually_added: bool
    present_in_database: bool


class CatalogSchemaEntryCreate(SQLModel):
    schema_name: str
    table_name: Optional[str] = None
    column_name: Optional[str] = None

    @field_validator("schema_name", "table_name", "column_name", mode="before")
    @classmethod
    def empty_str_to_none(cls, value):
        if isinstance(value, str):
            value = value.strip()
            return value or None
        return value


class CatalogSchemaSyncRead(SQLModel):
    catalog_id: int
    schemas: int
    tables: int
    columns: int
    created_entries: int
    deleted_entries: int
    deleted_rules: int


class CatalogTablePreviewRequest(SQLModel):
    user_id: int
    schema_name: str
    table_name: str
    limit: int = 5

    @field_validator("schema_name", "table_name", mode="before")
    @classmethod
    def strip_required_names(cls, value):
        if isinstance(value, str):
            return value.strip()
        return value

    @field_validator("limit")
    @classmethod
    def clamp_limit(cls, value: int) -> int:
        return max(1, min(int(value), 50))


class CatalogPreviewDatasetRead(SQLModel):
    columns: list[str]
    rows: list[dict[str, Any]]
    error: Optional[str] = None


class CatalogTablePreviewRead(SQLModel):
    catalog_id: int
    schema_name: str
    table_name: str
    limit: int
    before_maskql: CatalogPreviewDatasetRead
    after_maskql: CatalogPreviewDatasetRead
